### PR TITLE
Add output mcp4728

### PR DESCRIPTION
**Pull request by brotherdust**
# What does this implement/fix? 

Adds MCP4728 output component. A quad-channel variant of the MCP4725.

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Other

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):** esphome/esphome-docs#1310

## Test Environment

- [X] ESP32
- [ ] ESP8266

## Example entry for `config.yaml`:

```yaml
   # Example configuration entry

    # Set a global i2c connection
    i2c:
      sda: 21
      scl: 22
      scan: True
      id: i2c_0 # Optional if only one I2C bus is used.

    # Initialize the MCP4728 "bus" component
    mcp4728:
      address: 0x60

    # Set the output with default (address: 0x60 / global i2c)
    output:
    - platform: mcp4728
      id: dac_A
      channel: A
    - platform: mcp4728
      id: dac_B
      channel: B
    - platform: mcp4728
      id: dac_C
      channel: C
    - platform: mcp4728
      id: dac_D
      channel: D

```

## Checklist:
  - [X] The code change is tested and works locally.
  - [ ] Tests have been added to verify that the new code works (under `tests/` folder).

I might need some assistance with the testing part.
  
If user exposed functionality or configuration variables are added/changed:
  - [X] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

### CODE REVIEW

File: esphome/components/mcp4728/__init__.py [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+    .extend(i2c.i2c_device_schema(0x60))
+)
+
+
+def to_code(config):
+    var = cg.new_Pvariable(config[CONF_ID])
+    yield cg.register_component(var, config)
+    yield i2c.register_i2c_device(var, config)
+
+    # cg.add_library("adafruit/Adafruit MCP4728", "1.0.7")
```
Comment (OttoWinter) [inline esphome/components/mcp4728/__init__.py]: No commented out code please :)

File: esphome/components/mcp4728/mcp4728.cpp [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+  output_buffer[0] = sequential_write_cmd;
+  // VREF PD1 PD0 Gx D11 D10 D9 D8 [A] D7 D6 D5 D4 D3 D2 D1 D0 [A]
+  new_value |= (new_vref << 15);
+  new_value |= (new_pd_mode << 13);
+  new_value |= (new_gain << 12);
+
+  output_buffer[1] = new_value >> 8;
+  output_buffer[2] = new_value & 0xFF;
+
+  if (!this->write_bytes_raw(output_buffer, 3)) {
```
Comment (OttoWinter) [inline esphome/components/mcp4728/mcp4728.cpp]:
```suggestion
  return this->write_bytes_raw(output_buffer, 3);
```

See the clang-tidy message

File: esphome/components/mcp4728/mcp4728.cpp [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+      break;
+    case 1:
+      channel_enum = MCP4728_CHANNEL_B;
+      break;
+    case 2:
+      channel_enum = MCP4728_CHANNEL_C;
+      break;
+    case 3:
+      channel_enum = MCP4728_CHANNEL_D;
+      break;
```
Comment (OttoWinter) [inline esphome/components/mcp4728/mcp4728.cpp:58]: add a `default` case here to make the linter happy.

File: esphome/components/mcp4728/mcp4728.h [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+extern const uint8_t MCP4728_MULTI_IR_CMD;
+extern const uint8_t MCP4728_MULTI_EEPROM_CMD;
+extern const uint8_t MCP4728_FAST_WRITE_CMD;
+
+typedef enum pd_mode {
+  MCP4728_PD_MODE_NORMAL,
+  MCP4728_PD_MODE_GND_1K,
+  MCP4728_PD_MODE_GND_100K,
+  MCP4728_PD_MODE_GND_500K
+} MCP4728_pd_mode_t;
```
Comment (OttoWinter) [inline esphome/components/mcp4728/mcp4728.h]:
```suggestion
enum MCP4728PDMode {
  MCP4728_PD_MODE_NORMAL,
  MCP4728_PD_MODE_GND_1K,
  MCP4728_PD_MODE_GND_100K,
  MCP4728_PD_MODE_GND_500K
};
```

Here we're in C++ and we don't need to use typedef enums :)

File: esphome/components/mcp4728/mcp4728.h [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+
+typedef enum gain { MCP4728_GAIN_1X, MCP4728_GAIN_2X } MCP4728_gain_t;
```
Comment (OttoWinter) [inline esphome/components/mcp4728/mcp4728.h]:
```suggestion
enum MCP4728Gain {
  MCP4728_GAIN_1X, 
  MCP4728_GAIN_2X
};
```

File: esphome/components/mcp4728/mcp4728.h [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+ public:
+  MCP4728OutputComponent() {}
+  MCP4728Channel *create_channel(uint8_t channel);
+
+  void setup() override;
+  float get_setup_priority() const override { return setup_priority::HARDWARE; }
+
+ protected:
+  friend MCP4728Channel;
+  bool set_channel_value_(MCP4728_channel_t channel, uint16_t new_value, MCP4728_vref_t new_vref,
```
Comment (OttoWinter) [inline esphome/components/mcp4728/mcp4728.h]:
```suggestion
  bool set_channel_value_(MCP4728ChannelType channel, uint16_t new_value, MCP4728VRef new_vref,
```

Same for other enums

File: esphome/components/mcp4728/output.py [hunk 7]
```diff
@@ -0,0 +1,25 @@
...
+DEPENDENCIES = ["mcp4728"]
+
+MCP4728Channel = mcp4728_ns.class_("MCP4728Channel", output.FloatOutput)
+CONF_MCP4728_ID = "mcp4728_id"
+
+CONFIG_SCHEMA = output.FLOAT_OUTPUT_SCHEMA.extend(
+    {
+        cv.Required(CONF_ID): cv.declare_id(MCP4728Channel),
+        cv.GenerateID(CONF_MCP4728_ID): cv.use_id(MCP4728Output),
+        cv.Required(CONF_CHANNEL): cv.enum({"A": 0, "B": 1, "C": 2, "D": 3}),
```
Comment (OttoWinter) [inline esphome/components/mcp4728/output.py]: This should instead use the enum values. See for example `ads1115/sensor.py`, you can it like this:

```python3
MCP4728ChannelType = mcp4728_ns.enum("MCP4728ChannelType")
CHANNELS = {
	"A": MCP4728ChannelType.MCP4728_CHANNEL_A,
    # ...
}
```

and in `cv.enum` you should set `upper=True` to allow lowercase channel names

**Review by OttoWinter — COMMENTED** [review]
Thanks! I left you some comments (mostly about code style), could you have a look?

**Comment by brotherdust** [comment]
Absolutely! I'll get it done in the next couple of day!

File: esphome/components/mcp4728/mcp4728.h (hunk 4 above)
Comment (brotherdust) [reply esphome/components/mcp4728/mcp4728.h]: Done

File: esphome/components/mcp4728/mcp4728.cpp [hunk 8]
```diff
@@ -0,0 +1,70 @@
...
+    default:
+      break;
```
Comment (OttoWinter) [inline esphome/components/mcp4728/mcp4728.cpp]:
```suggestion
      return;
```

Or something similar, see the error annotations in the files tab.

File: esphome/components/mcp4728/output.py [hunk 9]
```diff
@@ -0,0 +1,34 @@
+import esphome.config_validation as cv
+import esphome.codegen as cg
+from esphome.components import output, i2c
```
Comment (OttoWinter) [inline esphome/components/mcp4728/output.py]:
```suggestion
from esphome.components import output
```

File: esphome/components/mcp4728/output.py [hunk 10]
```diff
@@ -0,0 +1,34 @@
...
+from esphome.const import CONF_ID, CONF_CHANNEL
+from . import MCP4728Output, mcp4728_ns
+
+DEPENDENCIES = ["mcp4728"]
+
+MCP4728Channel = mcp4728_ns.class_("MCP4728Channel", output.FloatOutput)
+CONF_MCP4728_ID = "mcp4728_id"
+
+MCP4727ChannelID = mcp4728_ns.enum("MCP4728ChannelType")
```
Comment (OttoWinter) [inline esphome/components/mcp4728/output.py]:
```suggestion
MCP4727ChannelID = mcp4728_ns.enum("MCP4727ChannelID")
```

File: esphome/components/mcp4728/output.py [hunk 11]
```diff
@@ -0,0 +1,34 @@
...
+CONFIG_SCHEMA = output.FLOAT_OUTPUT_SCHEMA.extend(
+    {
+        cv.Required(CONF_ID): cv.declare_id(MCP4728Channel),
+        cv.GenerateID(CONF_MCP4728_ID): cv.use_id(MCP4728Output),
+        cv.Required(CONF_CHANNEL): cv.enum(CHANNELS, upper=True),
+    }
+)
+
+
+
```
Comment (OttoWinter) [inline esphome/components/mcp4728/output.py]:
```suggestion

```

See lint errors

File: esphome/components/mcp4728/__init__.py [hunk 12]
```diff
@@ -0,0 +1,27 @@
+import esphome.codegen as cg
+import esphome.config_validation as cv
+from esphome.components import i2c
+from esphome.const import CONF_ID
+
+DEPENDENCIES = ["i2c"]
+MULTI_CONF = True
+AUTO_LOAD = ["output"]
+
```
Comment (jesserockz) [inline esphome/components/mcp4728/__init__.py:10]: Please also add yourself as a CODEOWNER and run `script/build_codeowners.py`
```suggestion
CODEOWNERS = ["@brotherdust"]

```

File: esphome/components/mcp4728/__init__.py [hunk 13]
```diff
@@ -0,0 +1,28 @@
...
+from esphome.components import i2c
+from esphome.const import CONF_ID
+
+DEPENDENCIES = ["i2c"]
+MULTI_CONF = True
+AUTO_LOAD = ["output"]
+CODEOWNERS = ["@brotherdust"]
+
+mcp4728_ns = cg.esphome_ns.namespace("mcp4728")
+MCP4728Output = mcp4728_ns.class_("MCP4728OutputComponent", cg.Component, i2c.I2CDevice)
```
Comment (jesserockz) [inline esphome/components/mcp4728/__init__.py]: Just for less confusion, this should match the c++ class name
```suggestion
MCP4728OutputComponent = mcp4728_ns.class_("MCP4728OutputComponent", cg.Component, i2c.I2CDevice)
```

File: esphome/components/mcp4728/__init__.py [hunk 14]
```diff
@@ -0,0 +1,28 @@
...
+
+CONFIG_SCHEMA = (
+    cv.Schema(
+        {
+            cv.GenerateID(): cv.declare_id(MCP4728Output),
```
Comment (jesserockz) [inline esphome/components/mcp4728/__init__.py]:
```suggestion
            cv.GenerateID(): cv.declare_id(MCP4728OutputComponent),
```

**Comment by oxan** [comment]
@brotherdust Can you add this component to one of the YAML files in tests/? Other than that, this seems ready to go!

File: esphome/components/mcp4728/output.py [hunk 15]
```diff
@@ -0,0 +1,33 @@
+import esphome.config_validation as cv
+import esphome.codegen as cg
+from esphome.components import output
+from esphome.const import CONF_ID, CONF_CHANNEL
+from . import MCP4728Output, mcp4728_ns
```
Comment (jesserockz) [inline esphome/components/mcp4728/output.py]:
```suggestion
from . import MCP4728OutputComponent, mcp4728_ns
```

File: esphome/components/mcp4728/output.py [hunk 16]
```diff
@@ -0,0 +1,33 @@
...
+    "A": MCP4727ChannelID.MCP4728_CHANNEL_A,
+    "B": MCP4727ChannelID.MCP4728_CHANNEL_B,
+    "C": MCP4727ChannelID.MCP4728_CHANNEL_C,
+    "D": MCP4727ChannelID.MCP4728_CHANNEL_D,
+}
+
+CONFIG_SCHEMA = output.FLOAT_OUTPUT_SCHEMA.extend(
+    {
+        cv.Required(CONF_ID): cv.declare_id(MCP4728Channel),
+        cv.GenerateID(CONF_MCP4728_ID): cv.use_id(MCP4728Output),
```
Comment (jesserockz) [inline esphome/components/mcp4728/output.py]:
```suggestion
        cv.GenerateID(CONF_MCP4728_ID): cv.use_id(MCP4728OutputComponent),
```

**Pull request closed without merging**